### PR TITLE
cudatext: init at 1.111.0

### LINKED ISSUE #56339: Please package CudaText
Cross-platform text editor, written in Lazarus

> CudaText is like an open source Sublime Text. Its core is written in Object Pascal, which makes it fast and agile and its plug-in system ist based on Python, which makes it qickly extensible. The developers are very responsive and kind.

From: https://www.slant.co/topics/12/viewpoints/74/~best-programming-text-editors~cudatext

Repository: https://github.com/Alexey-T/CudaText
Mirror: http://uvviewsoft.com/cudatext/download.html

Development seems very active (last commit 15 minutes ago).

Might be a nice open source alternative to SublimeText.

### PR DESCRIPTION
**Pull request by sikmir**
###### Motivation for this change
**[CudaText](http://uvviewsoft.com/cudatext)** is a cross-platform text editor, written in Lazarus.
Resolves #56339.

###### Things done
- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by davidak — APPROVED** [review]
looks good
works basically

some things don't work like install plugins, scroll tab left/right buttons, maybe more
but it seems already usable

Screenshot from 2020-09-04 05-15-10

**Comment by sikmir** [comment]
FYI @Alexey-T

**Comment by Alexey-T** [comment]
>some things don't work like install plugins,

It works, after you configure the "pylib__linux" option
https://wiki.freepascal.org/CudaText#Python_on_Linux.2C_BSD.2C_Solaris

> scroll tab left/right buttons,

It works. on your screen, we have little count of tabs so they don't scroll.

**Comment by sikmir** [comment]
> It works, after you configure the "pylib__linux" option

Before build, I set full path to `libpython3.so` in https://github.com/Alexey-T/CudaText/blob/3cbe36fb8aa3ee67262fc6079335259c0589edeb/app/proc_globdata.pas#L976
Am I right that after that it's not needed to specify full path in `settings_default`?
```sh
$ grep "\"pylib__linux\":" result/share/cudatext/settings_default/default.json
  "pylib__linux": "libpython3.so",
```

**Comment by sikmir** [comment]
Steps to reproduce:
1. Plugins -> Addons Manager -> Install (Nix lexer) - "Package installed", no errors.
2. Open some `flake.nix` file, dialog "Lexer(s) for flake.nix", choose "Download lexer: Nix", no errors, nothing happens, no highlighting.

**Comment by sikmir** [comment]
I found the root cause, CudaText on first start create `~/.config/cudatext/data/lexlib` with 555 permissions, instead of 755.

**Comment by Alexey-T** [comment]
> > It works, after you configure the "pylib__linux" option
> 
> Before build, I set full path to `libpython3.so` in https://github.com/Alexey-T/CudaText/blob/3cbe36fb8aa3ee67262fc6079335259c0589edeb/app/proc_globdata.pas#L976
> Am I right that after that it's not needed to specify full path in `settings_default`?

IMO yes, it should work.

> ```shell
> $ grep "\"pylib__linux\":" result/share/cudatext/settings_default/default.json
>   "pylib__linux": "libpython3.so",

but default.json is not read by program, it's only for user reference.

**Comment by Alexey-T** [comment]
> I've just found the root cause, CudaText on first start copy data folder from nix store into `~/.config/cudatext/data` with 555 permissions, instead of 755.

Cud runs this:
```
  {$ifdef linux}
  if OpDirLocal<>OpDirExe then
    if IsDistroUpdateNeeded then
      if DirectoryExistsUTF8(OpDirPrecopy) then
        RunCommand('cp', ['-R', '-u', '-t',
          OpDirLocal,
          '/usr/share/cudatext/py',
          '/usr/share/cudatext/data',
          '/usr/share/cudatext/settings_default'
          ], S);
  {$endif}  
```
so permissions 555 is copy from your dir, which is 555.

**Comment by sikmir** [comment]
> so permissions 555 is copy from your dir, which is 555.

Indeed, folder in nix store have 555 permissions. IMHO, It would be good if CudaText would make no assumtion about source permissions and force 755 while copying. What do you think?

**Comment by Alexey-T** [comment]
Yes, agree. Will try to change that code part.

**Comment by Alexey-T** [comment]
Made this change. do you need new linux build?

**Comment by sikmir** [comment]
> Made this change.

Thanks!

> do you need new linux build?

No, nix package doesn't use binary CudaText packages, but build it from sources. So I can just apply your patch.

**Comment by sikmir** [comment]
[Nix lexer](https://github.com/Alexey-T/CudaText-lexers/tree/master/Nix) is now installed by default, just for convenience of Nix users.

**Comment by davidak** [comment]
I built it again, but can't see the changes. Plugins still don't appear and Nix is not known.

`davidak@gaming:~/code/nixpkgs]$ nix-shell -p nixpkgs-review --run "nixpkgs-review pr 97063"`

**Comment by Alexey-T** [comment]
Check that file Nix.lcf exists in the .config/cudatext/data/lexlib.

**Comment by Alexey-T** [comment]
Plugins don't appear? But python console panel input works OK? press Ctrl+tilde - input works ok?

**Comment by sikmir** [comment]
> I built it again, but can't see the changes. Plugins still don't appear and Nix is not known.
> 
> `davidak@gaming:~/code/nixpkgs]$ nix-shell -p nixpkgs-review --run "nixpkgs-review pr 97063"`

I've not commited yet, wait a bit...

**Comment by sikmir** [comment]
@davidak Check it now, it works for me.

**Comment by Alexey-T** [comment]
Just note that currrent  version is 1.111.0.

**Comment by sikmir** [comment]
> Just note that currrent version is 1.111.0.

I don't see 1.111.0 in https://github.com/Alexey-T/CudaText/tags.

**Comment by Alexey-T** [comment]
Ah, "tag" will be little later, in 1-2 days.

**Comment by sikmir** [comment]
**TODO** for future:
- [ ] build on darwin (requires [fpc](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/fpc/default.nix)/[lazarus](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/fpc/lazarus.nix) support on darwin)
- [ ] optional [build with qt5](https://wiki.freepascal.org/CudaText#Linux_Qt5_build) widgetset (requires [libqt5pas](https://wiki.freepascal.org/Qt5_Interface#libqt5pas) missed in nixpkgs)

**Comment by davidak** [comment]
It still don't work.

It is definitely build from a15ce2e and i have done a `nix-collect-garbage -d` to actually build it again.

```
[davidak@nixos:~]$ tree .config/cudatext/data/lexlib
.config/cudatext/data/lexlib
├── Assembly.cuda-lexmap
├── Assembly.lcf
├── Bash script.cuda-lexmap
├── Bash script.lcf
├── Batch files.cuda-lexmap
├── Batch files.lcf
├── C++.cuda-lexmap
├── C.cuda-lexmap
├── C++.lcf
├── C.lcf
├── CSS.cuda-lexmap
├── CSS.lcf
├── HTML.cuda-lexmap
├── HTML.lcf
├── Ini files.cuda-lexmap
├── Ini files.lcf
├── JavaScript.cuda-lexmap
├── JavaScript.lcf
├── JSDoc.cuda-lexmap
├── JSDoc.lcf
├── JSON.cuda-lexmap
├── JSON.lcf
├── Lua.cuda-lexmap
├── Lua.lcf
├── Markdown.cuda-lexmap
├── Markdown.lcf
├── PHP.cuda-lexmap
├── PHP_.cuda-lexmap
├── PHP.lcf
├── PHP_.lcf
├── Python.cuda-lexmap
├── Python.lcf
├── RegEx.cuda-lexmap
├── RegEx.lcf
├── reStructuredText.cuda-lexmap
├── reStructuredText.lcf
├── Search results.cuda-lexmap
├── Search results.lcf
├── VBScript.cuda-lexmap
├── VBScript.lcf
├── XML.cuda-lexmap
├── XML.lcf
├── YAML.cuda-lexmap
└── YAML.lcf

0 directories, 44 files
```

**Comment by sikmir** [comment]
> It still don't work.

Remove your old `~/.config/cudatext` folder.

**Comment by davidak** [comment]
Yeah, they had the wrong permission.

`rm: cannot remove '/home/davidak/.config/cudatext/py/cuda_tree_markdown/__init__.py': Permission denied`

Now it works.

Screenshot from 2020-09-05 18-41-54

Also tested on different system.

**Review by davidak — APPROVED** [review]
looks good
works

**Comment by sikmir** [comment]
> Ah, "tag" will be little later, in 1-2 days.

Well, I will update this PR with new version as it will be out.

**Comment by sikmir** [comment]
@Alexey-T Just in case, is there any repo where I can find CudaText translations other than https://sourceforge.net/projects/cudatext/files/addons/translations/? I can't find it in https://github.com/CudaText-addons as well. I want to allow users to optionally install more languages by default, like it's done for libreoffice https://github.com/NixOS/nixpkgs/blob/ccf01af387838dc451283e50a64dbe83b1899474/pkgs/applications/office/libreoffice/default.nix#L16.
The problem is that we need somehow versioned permanent links, but sourceforge files are not (after translations files update, links will be the same, but sha256 will be changed).

**Comment by sikmir** [comment]
@davidak I've added `libqt5pas`, so now CudaText can be built with qt5 widgetset as well. And I've switched default widgetset from gtk2 to qt5, it looks a bit better for me.

File: pkgs/applications/editors/cudatext/default.nix [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    sha256 = "0mlb41q2wjlpz6myp6rhrb8bfz19357kfvvr8nsxbgzc4hspmldv";
+  };
+
+  postPatch = ''
+    substituteInPlace app/proc_globdata.pas \
+      --replace "/usr/share/cudatext" "$out/share/cudatext" \
+      --replace "libpython3.so" "${python3}/lib/libpython3.so"
+
+    # Don't check for update
+    sed -i "2138imnuHelpCheckUpd.Enabled:=false;" app/formmain.pas
```
Comment (7c6f434c) [inline pkgs/applications/editors/cudatext/default.nix:56]: Isn't this just _too_ fragile? Maybe better to search for the function name and insert after it starts, or use a patch?

Comment (sikmir) [reply pkgs/applications/editors/cudatext/default.nix:56]: OK, fixed.

**Comment by davidak** [comment]
With the dark theme "sub", the menu text is now black on dark grey.

Screenshot from 2020-09-06 23-35-15

terminal output:

```
[nix-shell:~/.cache/nixpkgs-review/pr-97063-7]$ cudatext
QApplication: invalid style override 'adwaita' passed, ignoring it.
	Available styles: Breeze, Windows, Fusion
```

**Comment by sikmir** [comment]
Hmm, here is my (qt5, there is no errors in terminal):
scr1
Maybe it makes sense to switch default option back to gtk2.

**Comment by davidak** [comment]
I'm using the GTK-based [Pantheon desktop](https://elementary.io/open-source).

**Comment by 7c6f434c** [comment]
@davidak Maybe you need to install the Adwaita theme files?

@sikmir do you think this is ready to merge as it is (mention me once you have a follow-up PR with more improvements), or is it better to wait for some soon-to-happen debugging/polishing?

**Comment by sikmir** [comment]
> @sikmir do you think this is ready to merge as it is (mention me once you have a follow-up PR with more improvements), or is it better to wait for some soon-to-happen debugging/polishing?

1.111.0 is out meanwhile (fixes problem with permissions, so we can remove patching), so let me to update and I thnk we are ready to proceed.

**Comment by sikmir** [comment]
@7c6f434c Update to 1.111.0 is done and verified, feel free to proceed with merge.

**Pull request merged**